### ISSUE TITLE
Tweak typescript config: use Project References in local development

### ISSUE THREAD

**Issue by mitchhentgesspotify** [backstage/backstage#13947]
We can improve developer lint-iteration time by leveraging [TypeScript Project References](https://www.typescriptlang.org/docs/handbook/project-references.html)

## Feature Suggestion

The fastest way to have TypeScript lint a large repository like `backstage` is to have it pretend everything is in "one big project", and lint it all in one go.

However, when developing locally, TypeScript is run more frequently - sometimes even iteratively when working on a change whose types won't quite fit together :)
Even when "incremental", TypeScript does a _lot_ of work if it's scanning the entire repository to see what changes are needed.

If Project References are used, then TypeScript is much faster at "filtering out" already-compiled sub-projects, and only doing work on the changed pieces.

(I'd have to re-confirm my measurements, but) IIRC, the iteration time of a one-line change can be lowered from roughly ~10-15s to ~2s.

## Possible Implementation

This isn't as simple as just porting to Project References, as "full, clean compiles" would be slower, which is bad for CI (though, perhaps we can cache and re-use Project Reference artifacts between jobs? That's out-of-scope :wink: ).
So, we'd need to be clever here and have essentially a dual-config situation:
1. Have as much common config as possible defined in-repo as conventionally as possible
2. Ensure when when a developer does TypeScript linting, it defaults to Project-References-style linting
3. When run in CI, use `tsc` CLI parameters/environment variables/manually specified "override config files/etc to do a "single pass" lint.

Finally, there's a tradeoff here: for developers, the first time they lint their repo, it *will* take longer - roughly 2.5x longer than with the existing setup. However, I'm estimating that this cost (where the developer can multitask elsewhere - heck, they just waited for a `yarn install` likely, what's another couple minutes) is much smaller than the current iteration time cost.

-----

I'd be happy to push this forward sometime in the next few months when I have time.

**Comment by Rugvip** [comment]
👍

I wanna highlight another option here too, which is to do something similar to this, but for local dev:  https://github.com/backstage/backstage/blob/16889b6ea7e57c35fad5b652c42eab3aab1d85ce/scripts/api-extractor.ts#L268

It's a temporary TS config that focuses in on the part of the project, rather than including all packages

**Comment by mitchhentgesspotify** [comment]
That's a good alternative option, though cleanup hooks can be less dependable such as if a process ends unexpectedly, system shutdown, etc: causing these files to appear as extra changes in the `git` working area, which devs would have to be aware of.

The other downside of that approach is that, if dependent projects are used (but not modified), it has the same issue as doing a top-level lint in that more files are being verified than necessary.

**Comment by mitchhentgesspotify** [comment]
This is still valid, but needs some time to fully describe the alternative solution (which IIRC was something like "use project references for local dev, but 'single-project monolith builds' for CI, etc)

**Comment by mitchhentgesspotify** [comment]
I looked further into doing a proof-of-concept of this within `backstage-frontend`, with two findings:

1. There's dependencies between packages within the monorepo, and this must be reflected twice within each dependent package: once within `package.json` and once within `tsconfig.json`. Fortunately, there's a [`utils-typescript-references` tool](https://www.npmjs.com/package/@goldstack/utils-typescript-references) which we could employ in git commit hooks / CI to ensure that these remain up-to-date.
2. Project references do not allow circular dependencies, and unfortunately there's still many of those within internal Backstage. Those will need to be resolved first before we can experiment there.

**Comment by mitchhentgesspotify** [comment]
Bump.
